wimbd/es/__init__.py

```diff
@@ -406,3 +407,160 @@ def count_total_occurrences_of_unigrams(
+                                              ) -> Dict:
+    matched_query = []
+    if all_phrases:
+        which_bool = "must" if do_score else "filter"
+        minimum_should_match = None
+    else:
+        which_bool = "should"
+        minimum_should_match = 1
+        
+    for phrase in phrases:
+        query = {"match_phrase": {"text": {"query": phrase, "slop": slop}}}
+        matched_query.append(query)
+    
+    query = {
+        "bool": {which_bool: matched_query, "minimum_should_match": minimum_should_match}
+    }
+    return query
+
+
+def get_documents_containing_phrase_with_slop(
+    index: str,
+    phrases: Union[str, List[str]],
+    all_phrases: bool = False,
+    slop: int = 5,
+    num_documents: int = 10,
+    return_all_hits: bool = False,
+    sort_field: str = "date",
+    subset_filter: Optional[Dict[str, Any]] = None,
+    es: Optional[Elasticsearch] = None,
+) -> Generator[Dict, None, None]:
+    """
+    
+    """
+    es = es or es_init()
+
+    if isinstance(phrases, str):
+        phrases = [phrases]
+    query = _query_documents_contain_phrase_with_slop(phrases, slop, all_phrases)
+
+    if subset_filter:
+        if "filter" not in query["bool"]:
+            query["bool"]["filter"] = []
+        for filter_ in subset_filter:
+            query["bool"]["filter"].append({"term": filter_})
+
+    if return_all_hits:
+        sort = [{sort_field: "asc"}]
+        pit = es.open_point_in_time(index=index, keep_alive="1m")
+        results = es.search(index=index, query=query, size=num_documents, sort=sort)[
+            "hits"
+        ]["hits"]
+        yield from results
+        while len(results) > 0:
+            results = es.search(
+                index=index,
+                query=query,
+                size=num_documents,
+                sort=sort,
+                search_after=results[-1]["sort"],
+            )["hits"]["hits"]
+            yield from results
+        try:
+            es.close_point_in_time(id=pit["id"])
+        except NotFoundError:
+            pass
+    else:
+        yield from es.search(index=index, query=query, size=num_documents)["hits"][
+            "hits"
+        ]
+
+
+def count_documents_with_regex_slop(
+    index: str,
+    phrases: Union[str, Iterable[str]],
+    all_phrases: bool = False,
+    slop: int = 0,
+    subset_filter: Optional[Dict[str, Any]] = None,
+    exact_phrases: Optional[Union[str, Iterable[str]]] = None,
+    es: Optional[Elasticsearch] = None,
+) -> Dict[str, int]:
+    """
+    Count the number of documents matching each string with a specified slop.
+
+    :param index: Name of the index
+    :param phrases: A single string or a list of strings to be matched in the `text` field.
+    :param all_phrases: Whether the document should contain all phrases (AND clause) or any
+        of the phrases (OR clause).
+    :param slop: The number of words allowed between the tokens in the string.
```

Comment:
So in this case, we're looking for documents where the phrases "legal based" and "licenses" occur within 3 words of each other?